public/recorder.js

```diff
@@ -41,9 +56,186 @@ coverImg.appendChild(avatar);
+  // - No support for Promise version of decodeAudioData yet
+  const buffer = await new Promise(resolve => {
+    const reader = new FileReader();
+    reader.readAsArrayBuffer(recording);
+    reader.onloadend = () => {
+      resolve(reader.result);
+    }
+  });
+  const decodedAudio = await new Promise((res, rej) => audioContext.decodeAudioData(buffer, res, rej));
+  startTime = audioContext.currentTime;
+  playbackNode = audioContext.createBufferSource();
+  playbackNode.buffer = decodedAudio;
+  playbackNode.start(startTime);
+  playbackNode.connect(audioContext.destination);
+  playbackNode.addEventListener("ended", onEndedHandler);
+}
+
+function startKaraoke({ wait } = { wait: 0 }) {
+  let lastSegment = -1;
+  let lastLang = null;
+
+  function scheduleNextCheck() {
+    karaokeTimeout = setTimeout(karaokeCheck, 100);
+  }
+
+  function karaokeCheck() {
+    const currentTime = audioContext.currentTime - startTime - wait;
+    const segment = Math.floor(currentTime / segmentDuration);
+    const lang = lyrics[selectedLang] ? selectedLang : "en";
+
+    if ((segment < 0) || ((segment === lastSegment) && (lang === lastLang))) {
+      scheduleNextCheck();
+      return;
+    }
+    lastSegment = segment;
+    lastLang = lang;
+
+    const line = lyrics[lang][segment];
+    const nextLine = lyrics[lang][segment + 1] || false;
+    if (!line) {
+      karaoke.textContent = "";
+      scheduleNextCheck();
+      return;
+    }
+    const current = document.createElement("span");
+    let next;
+    if (nextLine) {
+      next = document.createElement("p");
+      next.className = "next";
+    }
+    if (line.lang) {
+      current.lang = line.lang;
+      current.textContent = line.string
+    } else {
+      current.textContent = line;
+    }
+    if (nextLine) {
+      if (line.lang) {
+        next.lang = nextLine.lang;
+        next.textContent = nextLine.string
+      } else {
+        next.textContent = nextLine;
+      }
+    }
+    karaoke.innerHTML = "";
+    karaoke.appendChild(current);
+    if (nextLine) {
+      karaoke.appendChild(next);
+    }
+
+    scheduleNextCheck();
+  }
+
+  stopKaraoke();
+  scheduleNextCheck();
+}
+
+function stopKaraoke() {
+  if (karaokeTimeout) {
+    clearTimeout(karaokeTimeout);
+    karaokeTimeout = null;
+  }
+}
+
+
 (async function() {
+  // Safari compat notes:
+  // - No support for Promise version of decodeAudioData yet
+  // - Call to suspend does not seem to set the internal "suspended by user"
+  // flag, see workaround in recordBtn click event handler.
+  await audioContext.suspend();
+  startTime = audioContext.currentTime;
+  toneDecodedAudio = await fetch(`audios/tone.${audioFormat}`)
+    .then(r => r.arrayBuffer())
+    .then(buffer => new Promise((res, rej) => audioContext.decodeAudioData(buffer, res, rej)));
+  refDecodedAudio = await fetch(`audios/reference.${audioFormat}`)
+    .then(r => r.arrayBuffer())
+    .then(buffer => {
+      refBlob = new Blob([buffer]);
```

Comment:
what is `refBlob` used for?

if nothing, the chain r=>r.arrayBuffer() → decodedAudio might be factored out for readability

Reply:
It used to have a purpose. Now, if only I remembered which one ;) I'll drop that.